R/api.R

```diff
@@ -0,0 +1,1137 @@
+                                                 max_delay_seconds = 300,
+                                                 timeout_seconds = 10800,
+                                                 verbose = TRUE,
+                                                 api_key = Sys.getenv("IPUMS_API_KEY")) {
+
+  stopifnot(is.numeric(initial_delay_seconds))
+  stopifnot(is.numeric(max_delay_seconds))
+  stopifnot(is.null(timeout_seconds) || is.numeric(timeout_seconds))
+
+  current_delay <- initial_delay_seconds
+  is_timed_out <- FALSE
+  is_finished <- FALSE
+  is_error_state <- FALSE
+  err_message <- "Unknown Error"
+
+  wait_start <- Sys.time()
+
+  while (!is_timed_out && !is_finished && !is_error_state) {
+    if (current_delay > 0) {
+      if (verbose) {
+        message(paste("Waiting", current_delay, "seconds..."))
+      }
+
+      Sys.sleep(current_delay)
+    }
+
+    if (verbose) {
+      message("Checking extract status...")
+    }
+    extract <- ipums_api_get_extract(collection, number, api_key)
+
+    is_downloadable <- ipums_api_extract_downloadable(extract)
+
+    is_failed <- !(is_downloadable ||
+                     extract$status %in% c("queued", "started", "produced"))
+    is_timed_out <- !is.null(timeout_seconds) &&
+      as.numeric(Sys.time() - wait_start, units = "secs") > timeout_seconds
+
+
+    if (is_downloadable) {
+      if (verbose) {
+        message("Extract ready to download")
+      }
+      is_finished <- TRUE
+    } else if (is_failed) {
+      err_message <- "Extract has finished, but is not in a downloadable state"
+      is_error_state <- TRUE
+    } else if (is_timed_out) {
+      err_message <- "Max timeout elapsed"
+    }
+
+    if (current_delay == 0) {
+      current_delay <- 2
+    } else {
+      current_delay <- min(c(current_delay * 2, max_delay_seconds))
+    }
+  }
+
+  if (!is_finished) {
+    stop(err_message)
+  }
+
+  extract
+}
+
+
+# > Check if downloadable ----
+
+#' Is the extract ready to download?
+#'
+#' This function checks whether the given extract is ready to download,
+#' returning TRUE for extracts that are ready and FALSE for those that are not.
+#'
+#' @param object Either an object of class \code{ipums_extract}, the name of an IPUMS
+#'   data collection as a length one character vector, or a data.frame with an
+#'   extract definition in each row.
+#' @inheritParams ipums_api_get_extract
+#' @param extract_df A data.frame in which each row contains the definition of
+#'   an IPUMS extract.
+#'
+#' @return A logical vector. If the first argument is an object of class
+#'   \code{ipums_extract} or the name of a data collection, the function will
+#'   return a length-one logical vector. If the first
+#'   argument is a \code{data.frame}, the function will return a logical vector
+#'   with length equal to the number of rows of the \code{data.frame}.
+#' @export
+ipums_api_extract_downloadable <- function(object, ...) {
```

Comment:
I added some more discussion of the various values that the "status" of an extract can take, and the distinction between "status" and whether an extract is ready to download, to the documentation for `is_extract_ready()` and to the vignette.

Reply:
I'm going to resolve this for now so that I am ready to merge this pull request in the near future, but we can continue this discussion in the form of an issue if necessary.